certs/certs.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	if err := verifier.VerifyAggregate(signedBytes, cert.Signature, signers); err != nil {
+		return xerrors.Errorf("invalid signature on finality certificate for instance %d: %w", cert.GPBFTInstance, err)
+	}
+	return nil
+}
+
+// MakePowerTableDiff create a power table diff between the two given power tables. It makes no
+// assumptions about order, but does assume that the power table entries are unique. The returned
+// diff is sorted by participant ID ascending.
+func MakePowerTableDiff(oldPowerTable, newPowerTable gpbft.PowerEntries) []PowerTableDelta {
+	oldPowerMap := make(map[gpbft.ActorID]*gpbft.PowerEntry, len(oldPowerTable))
+	for i := range oldPowerTable {
+		e := &oldPowerTable[i]
+		oldPowerMap[e.ID] = e
+	}
+
+	var diff []PowerTableDelta
+	for i := range newPowerTable {
+		newEntry := &newPowerTable[i]
+		delta := PowerTableDelta{ParticipantID: newEntry.ID}
+		if oldEntry, ok := oldPowerMap[newEntry.ID]; ok {
+			delete(oldPowerMap, newEntry.ID)
+			delta.PowerDelta = new(gpbft.StoragePower).Sub(newEntry.Power, oldEntry.Power)
+			if !bytes.Equal(newEntry.PubKey, oldEntry.PubKey) {
+				delta.SigningKey = newEntry.PubKey
+			}
+			if delta.IsZero() {
+				continue
+			}
+		} else {
+			delta.PowerDelta = newEntry.Power
+			delta.SigningKey = newEntry.PubKey
+		}
+		diff = append(diff, delta)
+	}
+	for _, e := range oldPowerMap {
+		diff = append(diff, PowerTableDelta{
+			ParticipantID: e.ID,
+			PowerDelta:    new(gpbft.StoragePower).Neg(e.Power),
+		})
+	}
+	slices.SortFunc(diff, func(a, b PowerTableDelta) int {
+		return cmp.Compare(a.ParticipantID, b.ParticipantID)
+	})
+	return diff
+}
+
+// Apply a power table diff to the passed power table.
+//
+// - The delta must be sorted by participant ID, ascending.
+// - The returned power table is sorted by power, descending.
+func ApplyPowerTableDiff(prevPowerTable gpbft.PowerEntries, delta []PowerTableDelta) (gpbft.PowerEntries, error) {
+	deltaMap := make(map[gpbft.ActorID]*PowerTableDelta, len(delta))
+	var lastActorId gpbft.ActorID
+	for i := range delta {
+		d := &delta[i]
+
+		// We assert this to make sure the finality certificate has a consistent power-table
+		// diff.
+		if i > 0 && d.ParticipantID <= lastActorId {
+			return nil, xerrors.Errorf("power table delta not sorted by participant ID")
+		}
+
+		deltaMap[d.ParticipantID] = d
+		lastActorId = d.ParticipantID
+	}
+
+	// NOTE: we don't check if the new power is negative or anything like that because
+	// we're just going to serialize to CBOR and validate that the hashes map. So the
+	// only sanity check we need is the above "sort" check.
+	newPowerTable := make(gpbft.PowerEntries, 0, len(prevPowerTable)+len(deltaMap))
+	for _, pe := range prevPowerTable {
+		if diff, ok := deltaMap[pe.ID]; ok {
+			delete(deltaMap, pe.ID)
+
+			if diff.PowerDelta.Sign() != 0 {
+				pe.Power = new(gpbft.StoragePower).Add(diff.PowerDelta, pe.Power)
+				if pe.Power.Sign() == 0 {
+					continue
```

Comment:
The interpretation of a deletion here is subtle, worth a comment.

Reply:
Fixed in #291.